### PR TITLE
Feat/password reset

### PR DESCRIPTION
**Pull request by mingard**

### CODE REVIEW

File: app/lib/controllers/session.js [hunk 1]
```diff
@@ -112,6 +154,94 @@ Session.prototype.put = function (req, res, next) {
   return next()
 }
 
+/**
+ * Reset
```
Comment (eduardoboucas) [inline app/lib/controllers/session.js]: Why the repetition in the JSdoc comments (e.g. `Reset` + `Reset password`)?

File: app/lib/controllers/session.js [hunk 2]
```diff
@@ -112,6 +154,94 @@ Session.prototype.put = function (req, res, next) {
...
+Session.prototype.reset = function (req, res, next) {
+  res.header('Content-Type', 'application/json')
+
+  let authAPI = config.get('auth')
+
+  // No authentication.
+  if (!authAPI.enabled) {
+    res.write(JSON.stringify({error: 'AUTH_DISABLED'}))
+    res.end()
+    return next()
```
Comment (eduardoboucas) [inline app/lib/controllers/session.js]: Could you add an empty line before the `return`?

File: app/lib/controllers/session.js [hunk 3]
```diff
@@ -112,6 +154,94 @@ Session.prototype.put = function (req, res, next) {
...
+  }
+
+  // No authentication.
+  if (!authAPI.enabled) {
+    res.write(JSON.stringify({error: 'AUTH_DISABLED'}))
+    res.end()
+    return next()
```
Comment (eduardoboucas) [inline app/lib/controllers/session.js]: Could you add an empty line before the `return`?

File: app/lib/controllers/session.js [hunk 4]
```diff
@@ -112,6 +154,94 @@ Session.prototype.put = function (req, res, next) {
...
+Session.prototype.resetToken = function (req, res, next) {
+  res.header('Content-Type', 'application/json')
+
+  let authAPI = config.get('auth')
+
+  // No authentication.
+  if (!authAPI.enabled) {
+    res.write(JSON.stringify({error: 'AUTH_DISABLED'}))
+    res.end()
+    return next()
```
Comment (eduardoboucas) [inline app/lib/controllers/session.js]: Could you add an empty line before the `return`?

File: frontend/actions/actionTypes.js [hunk 5]
```diff
@@ -18,6 +18,8 @@ export const SET_REMOTE_USER = 'SET_REMOTE_USER'
 export const SET_USER_STATUS = 'SET_USER_STATUS'
 export const SIGN_OUT = 'SIGN_OUT'
 export const UPDATE_LOCAL_USER = 'UPDATE_LOCAL_USER'
+export const REQUEST_PASSWORD_RESET = 'REQUEST_PASSWORD_RESET'
```
Comment (eduardoboucas) [inline frontend/actions/actionTypes.js]: Can you maintain alphabetical order, please?

File: frontend/actions/userActions.js [hunk 6]
```diff
@@ -23,6 +23,54 @@ export function registerFailedSignInAttempt () {
...
+    success: response.success,
+    type: Types.REQUEST_PASSWORD_RESET_SUCCESS
+  }
+}
+
+export function requestPasswordReset (resetEmail) {
+  return (dispatch, getState) => {
+    return runSessionQuery({path: '/session/reset-token', payload: {email: resetEmail}})
+      .then(response => {
+        const resetExpiresAt = response.expiresAt
```
Comment (eduardoboucas) [inline frontend/actions/userActions.js]: Do we need this variable here?

File: frontend/actions/userActions.js [hunk 7]
```diff
@@ -37,20 +85,32 @@ function runSessionQuery ({
   }
 
   if (payload) {
+    // Force POST method of there is a payload
```
Comment (eduardoboucas) [inline frontend/actions/userActions.js]: `of`-> `if`

File: frontend/components/Button/Button.jsx [hunk 8]
```diff
@@ -56,6 +56,11 @@ export default class Button extends Component {
     inGroup: proptypes.oneOf(['left', 'middle', 'right']),
 
     /**
+     * Button label.
+     */
+    label: proptypes.string,
```
Comment (eduardoboucas) [inline frontend/components/Button/Button.jsx]: What is this new prop for?

File: frontend/components/TextInput/TextInput.jsx [hunk 9]
```diff
@@ -155,7 +160,10 @@ export default class TextInput extends Component {
   }
 
   handleChange(event) {
-    const {onChange, onKeyUp} = this.props
+    const {onChange, onKeyUp, validation} = this.props
+
+    // Attach validation outcome to event.
+    event.isValid = validation? validation(event.target.value): true
```
Comment (eduardoboucas) [inline frontend/components/TextInput/TextInput.jsx]: Missing a space before the ternary operator:

```js
event.isValid = validation ? validation(event.target.value): true
```

File: frontend/containers/SignIn/SignIn.jsx [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+     /**
+      * Sign in token.
+      */
+     token: proptypes.string
+  }
+
+  constructor(props) {
+    super(props)
+
+    // Is there is a token in the URL.
```
Comment (eduardoboucas) [inline frontend/containers/SignIn/SignIn.jsx]: What does this comment mean?

File: frontend/containers/SignIn/SignIn.jsx [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+                <Banner>{error}</Banner>
+              }
+
+              <div class={styles.inputs}>
+                {!isPasswordReset && (
+                  <div class={styles.input}>
+                    <Label label="Email">
+                      <TextInput
+                        placeholder="Your email address"
+                        validation={this.validation.email}
```
Comment (eduardoboucas) [inline frontend/containers/SignIn/SignIn.jsx]: Can we maintain alphabetical order of props?

File: frontend/containers/SignIn/SignIn.jsx [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+                        value={this.state.email}
+                      />
+                    </Label>
+                  </div>
+                )}
+
+                <div class={styles.input}>
+                  <Label label={isPasswordReset ? 'New Password' : 'Password'}>
+                    <TextInput
+                      type="password"
```
Comment (eduardoboucas) [inline frontend/containers/SignIn/SignIn.jsx]: Can we maintain alphabetical order of props?

File: frontend/containers/SignIn/SignIn.jsx [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+                    </Label>
+                  </div>
+                )}
+              </div>
+
+              <Button
+                accent="system"
+                disabled={hasConnectionIssues || (!formDataIsValid && !isPasswordReset)}
+                type="submit"
+                label={isPasswordReset ? 'Reset password': 'Sign In'}
```
Comment (eduardoboucas) [inline frontend/containers/SignIn/SignIn.jsx]: This goes back to a previous comment I added, but I still don't get why we need a `label` prop in `<Button>`. Why not just:

```jsx
<Button
  accent="system"
  disabled={hasConnectionIssues || (!formDataIsValid && !isPasswordReset)}
  type="submit"
>{isPasswordReset ? 'Reset password': 'Sign In'}</Button>
```

File: frontend/containers/SignIn/SignIn.jsx [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+    this.setState({
+      [name]: event.target.value,
+      formDataIsValid: event.isValid
+    })
+  }
+
+  handleSignIn(event) {
+    const {actions, token} = this.props
+    const {email, password, isPasswordReset, passwordConfirm} = this.state
+    if (isPasswordReset) {
```
Comment (eduardoboucas) [inline frontend/containers/SignIn/SignIn.jsx]: Could you add an empty line after the declarations block?

File: frontend/lib/constants.js [hunk 15]
```diff
@@ -22,3 +22,4 @@ export const STATUS_LOADING = 'STATUS_LOADING'
 export const STATUS_NOT_FOUND = 'STATUS_NOT_FOUND'
 export const STATUS_SAVING = 'STATUS_SAVING'
 export const STATUS_LOADED = 'STATUS_LOADED'
+export const PASSWORD_MISMATCH = 'PASSWORD_MISMATCH'
```
Comment (eduardoboucas) [inline frontend/lib/constants.js]: Could we maintain alphabetical order?

File: frontend/containers/SignIn/SignIn.jsx (hunk 12 above)
Comment (mingard) [inline frontend/containers/SignIn/SignIn.jsx]: Can we lint that?

Comment (eduardoboucas) [inline frontend/containers/SignIn/SignIn.jsx]: I'm not sure, I don't think we have any JSX-specific lint rules. I can find out if there's something we can use.

File: app/lib/controllers/session.js (hunk 1 above)
Comment (mingard) [inline app/lib/controllers/session.js]: Method title, method description

File: frontend/actions/userActions.js (hunk 6 above)
Comment (mingard) [inline frontend/actions/userActions.js]: Yes, just putting some changes in place to use this. I'll add notes

Comment (mingard) [inline frontend/actions/userActions.js]: Actually, I'll remove it for now

File: frontend/components/Button/Button.jsx (hunk 8 above)
Comment (mingard) [inline frontend/components/Button/Button.jsx]: I'm about to change this

File: app/lib/controllers/session.js (hunk 1 above)
Comment (eduardoboucas) [inline app/lib/controllers/session.js]: Why do we need method title? We haven't been using that and I can't find an example on [jsdoc.org](http://usejsdoc.org/index.html) with it.

File: frontend/components/Button/Button.jsx (hunk 8 above)
Comment (eduardoboucas) [inline frontend/components/Button/Button.jsx]: Okay, let me know once it's ready for review.

**Pull request merged**